rllib/examples/autoregressive_action_dist.py

```diff
@@ -0,0 +1,212 @@
+
+import gym
+from gym.spaces import Discrete, Tuple
+import argparse
+import random
+
+import ray
+from ray import tune
+from ray.rllib.models import ModelCatalog
+from ray.rllib.models.tf.tf_action_dist import Categorical, ActionDistribution
+from ray.rllib.models.tf.misc import normc_initializer
+from ray.rllib.models.tf.tf_modelv2 import TFModelV2
+from ray.rllib.policy.policy import TupleActions
+from ray.rllib.utils import try_import_tf
+
+tf = try_import_tf()
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--run", type=str, default="PPO")  # try PG, PPO, IMPALA
+parser.add_argument("--stop", type=int, default=200)
+
+
+class CorrelatedActionsEnv(gym.Env):
+    """Simple env in which the policy has to emit a tuple of equal actions.
+
+    The best score would be ~200 reward."""
+
+    def __init__(self, _):
+        self.observation_space = Discrete(2)
+        self.action_space = Tuple([Discrete(2), Discrete(2)])
+
+    def reset(self):
+        self.t = 0
+        self.last = random.choice([0, 1])
+        return self.last
+
+    def step(self, action):
+        self.t += 1
+        a1, a2 = action
+        reward = 0
+        if a1 == self.last:
+            reward += 5
+        # encourage correlation between a1 and a2
+        if a1 == a2:
+            reward += 5
+        done = self.t > 20
+        self.last = random.choice([0, 1])
+        return self.last, reward, done, {}
+
+
+class BinaryAutoregressiveOutput(ActionDistribution):
+    """Action distribution P(a1, a2) = P(a1) * P(a2 | a1)"""
+
+    @staticmethod
+    def required_model_output_shape(self, model_config):
+        return 16  # controls model output feature vector size
+
+    def sample(self):
+        # first, sample a1
+        a1_dist = self._a1_distribution()
+        a1 = a1_dist.sample()
+
+        # sample a2 conditioned on a1
+        a2_dist = self._a2_distribution(a1)
+        a2 = a2_dist.sample()
+        self._action_logp = a1_dist.logp(a1) + a2_dist.logp(a2)
+
+        # return the action tuple
+        return TupleActions([a1, a2])
+
+    def logp(self, actions):
+        a1, a2 = actions[:, 0], actions[:, 1]
+        a1_vec = tf.expand_dims(tf.cast(a1, tf.float32), 1)
+        a1_logits, a2_logits = self.model.action_model([self.inputs, a1_vec])
+        return (
+            Categorical(a1_logits).logp(a1) + Categorical(a2_logits).logp(a2))
+
+    def sampled_action_logp(self):
+        return tf.exp(self._action_logp)
+
+    def entropy(self):
+        a1_dist = self._a1_distribution()
+        a2_dist = self._a2_distribution(a1_dist.sample())
+        return a1_dist.entropy() + a2_dist.entropy()
+
+    def kl(self, other):
+        # TODO: implement this properly
+        return tf.zeros_like(self.entropy())
+
+    def _a1_distribution(self):
+        BATCH = tf.shape(self.inputs)[0]
+        a1_logits, _ = self.model.action_model(
+            [self.inputs, tf.zeros((BATCH, 1))])
+        a1_dist = Categorical(a1_logits)
```

Comment:
would this have issues? Like somehow adding nodes to the tf graph over and over?

Reply:
It's fine since this entire thing is only called once in graph mode.